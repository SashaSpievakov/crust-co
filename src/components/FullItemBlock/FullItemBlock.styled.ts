import styled from 'styled-components';
import { Container } from '../../styles/Base.styled';

const Item = styled(Container)`
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 20px;

  @media screen and (max-width: 600px) {
    gap: 15px;
  }

  @media screen and (max-width: 400px) {
    gap: 10px;
  }
`;

const Image = styled.img`
  margin-top: 60px;
  margin-bottom: 40px;
  border-radius: 20px;

  @media screen and (max-width: 600px) {
    width: 300px;
    height: 300px;
    margin-top: 50px;
    margin-bottom: 10px;
  }

  @media screen and (max-width: 400px) {
    width: 200px;
    height: 200px;
    margin-top: 20px;
    margin-bottom: 0;
  }
`;

const Title = styled.h2`
  font-size: 45px;
  font-weight: 700;
  text-align: center;

  @media screen and (max-width: 600px) {
    font-size: 35px;
    margin-bottom: 0;
  }

  @media screen and (max-width: 400px) {
    line-height: 35px;
  }
`;

const Paragraph = styled.p`
  font-size: 19px;
  max-width: 600px;
  text-align: center;
  margin-bottom: 15px;

  @media screen and (max-width: 550px) {
    font-size: 14px;
  }
`;

const Price = styled.h3`
  font-size: 45px;

  @media screen and (max-width: 500px) {
    font-size: 40px;
  }
`;

export { Item, Image, Title, Paragraph, Price };
