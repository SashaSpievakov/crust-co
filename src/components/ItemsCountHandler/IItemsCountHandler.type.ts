export interface IItemsCountHandler {
  id: string;
  name: string;
  price: number;
  activeSize: number;
  activeType: number;
  typeNames: string[];
  isFullScreen?: boolean;
}
