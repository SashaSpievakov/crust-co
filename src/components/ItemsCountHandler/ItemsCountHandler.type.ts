export interface ItemsCountHandlerType {
  id: string;
  name: string;
  price: number;
  activeSize: number;
  activeType: number;
  typeNames: string[];
  isFullScreen?: boolean;
}
