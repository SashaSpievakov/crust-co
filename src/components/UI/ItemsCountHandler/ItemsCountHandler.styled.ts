import styled, { css } from 'styled-components';
import { HiMinus, HiPlus } from 'react-icons/hi';
import { BsPlusLg } from 'react-icons/bs';

interface CounterProps {
  isFullScreen?: boolean;
}

interface MinusProps {
  disabled: boolean;
}

const mixinButton = css`
  font-size: 27px;
  color: ${({ theme }) => theme.colors.secondary};
  user-select: none;
  cursor: pointer;

  &:active {
    transform: translateY(1px);
  }
`;

const Counter = styled.section<CounterProps>`
  display: flex;
  align-items: center;
  min-width: 100px;
  padding: 8.25px 0;

  svg {
    font-size: ${({ isFullScreen }) => (isFullScreen ? '34px' : '27px')};
  }
`;

const Minus = styled(HiMinus)<MinusProps>`
  ${mixinButton}
  margin-right: 15px;

  ${({ disabled }) =>
    disabled &&
    css`
      color: ${({ theme }) => theme.colors.disabled};
      cursor: auto;
    `}
`;

const Plus = styled(HiPlus)`
  ${mixinButton}
`;

const Count = styled.span<CounterProps>`
  font-size: ${({ isFullScreen }) => (isFullScreen ? '30px' : '22px')};
  margin-right: 15px;
`;

const MainPlus = styled(BsPlusLg)`
  font-size: 17px;
  padding-bottom: 2px;
  margin-right: 5px;
`;

export { Counter, Minus, Plus, Count, MainPlus };
