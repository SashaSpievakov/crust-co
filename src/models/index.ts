export type { ICartItem } from './ICartItem';
export type { IPizzaItem } from './IPizzaItem';
