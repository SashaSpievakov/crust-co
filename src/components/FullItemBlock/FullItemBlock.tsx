import { useState } from 'react';

import { IPizzaItem } from '../../models';
import ItemsCountHandler from '../ItemsCountHandler/ItemsCountHandler';
import Selector from '../Selector/Selector';
import { ButtonLink } from '../UI';
import { Image, Item, Paragraph, Price, Title } from './FullItemBlock.styled';

interface FullItemProps {
  item: IPizzaItem;
}

export const typeNames: string[] = ['traditional', 'thin'];

const FullItemBlock = ({ item }: FullItemProps) => {
  const [activePrice, setActivePrice] = useState<number>(item.price);
  const [activeSize, setActiveSize] = useState<number>(12);
  const [activeType, setActiveType] = useState<number>(0);

  return (
    <Item data-testid="itemPage">
      <picture>
        <source
          srcSet={`../assets/img/pizza${item.id}.webp`}
          type="image/webp"
        />
        <source srcSet={`../assets/img/pizza${item.id}.png`} type="image/png" />
        <Image
          src={`../assets/img/pizza${item.id}.png`}
          alt={`${item.name} pizza`}
          width="450"
          height="450"
        />
      </picture>
      <Title>{item.name}</Title>
      <Paragraph>{item.description}</Paragraph>
      <Selector
        price={activePrice}
        sizes={item.sizes}
        types={item.types}
        activeSize={activeSize}
        activeType={activeType}
        setActivePrice={setActivePrice}
        setActiveSize={setActiveSize}
        setActiveType={setActiveType}
        typeNames={typeNames}
        isFullScreen
      />
      <Price>{activePrice}$</Price>
      <ItemsCountHandler
        id={item.id}
        name={item.name}
        price={activePrice}
        activeSize={activeSize}
        activeType={activeType}
        typeNames={typeNames}
        isFullScreen
      />
      <ButtonLink link="/">Go Back</ButtonLink>
    </Item>
  );
};

export default FullItemBlock;
