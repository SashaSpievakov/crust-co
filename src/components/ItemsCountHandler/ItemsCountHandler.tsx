import { FC, KeyboardEvent } from 'react';

import { useAppDispatch, useAppSelector } from '../../hooks';
import { ICartItem } from '../../models';
import {
  addItem,
  removeItem,
  selectCurrentItemCount,
  selectItemsCount,
} from '../../store';
import { ButtonTertiary } from '../../styles/Buttons.styled';
import {
  Count,
  Counter,
  MainPlus,
  Minus,
  Plus,
} from './ItemsCountHandler.styled';
import { ItemsCountHandlerType } from './ItemsCountHandler.type';

export const ItemsCountHandler: FC<ItemsCountHandlerType> = ({
  id,
  name,
  price,
  activeSize,
  activeType,
  typeNames,
  isFullScreen = false,
}) => {
  const dispatch = useAppDispatch();
  const namesCount = useAppSelector(selectItemsCount(id));
  const currentItemCount = useAppSelector(
    selectCurrentItemCount(name, activeSize, typeNames[activeType]),
  );

  const onClickAdd = () => {
    const item: ICartItem = {
      id,
      name,
      price,
      size: activeSize,
      type: typeNames[activeType],
      count: 0,
    };

    dispatch(addItem(item));
  };

  const onClickRemove = () => {
    if (currentItemCount < 1) return;

    const item: ICartItem = {
      id,
      name,
      price,
      size: activeSize,
      type: typeNames[activeType],
      count: 0,
    };

    dispatch(removeItem(item));
  };

  const handleAddKeyDownClick = (e: KeyboardEvent<SVGElement>) => {
    if (e.code === 'Enter') onClickAdd();
  };

  const handleRemoveKeyDownClick = (e: KeyboardEvent<SVGElement>) => {
    if (e.code === 'Enter') onClickRemove();
  };

  return namesCount ? (
    <Counter isFullScreen={isFullScreen}>
      <Minus
        onClick={onClickRemove}
        onKeyDown={(e) => handleRemoveKeyDownClick(e)}
        tabIndex={currentItemCount < 1 ? -1 : 0}
        disabled={currentItemCount < 1}
        data-testid="itemsHandlerMinus"
      />
      <Count isFullScreen={isFullScreen}>{namesCount}</Count>
      <Plus
        onClick={onClickAdd}
        onKeyDown={(e) => handleAddKeyDownClick(e)}
        tabIndex={0}
        data-testid="itemsHandlerPlus"
      />
    </Counter>
  ) : (
    <ButtonTertiary onClick={onClickAdd} isFullScreen={isFullScreen}>
      <MainPlus />
      <span>Add</span>
    </ButtonTertiary>
  );
};
