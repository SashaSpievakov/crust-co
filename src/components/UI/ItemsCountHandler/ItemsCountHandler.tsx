import { KeyboardEvent } from 'react';

import { useAppDispatch, useAppSelector } from '../../../hooks/reduxHooks';
import { ICartItem } from '../../../models/ICartItem';
import { IItemsCountHandler } from '../../../models/IItemsCountHandler';
import {
  addItem,
  removeItem,
} from '../../../store/slices/cart/reducer/cartReducer';
import { selectItemsCount } from '../../../store/slices/cart/selectors/selectItemsCount/selectItemsCount';
import { selectCurrentItemCount } from '../../../store/slices/cart/selectors/selectCurrentItemCount/selectCurrentItemCount';
import { ButtonAdd } from '../../../styles/Buttons.styled';
import {
  Counter,
  Minus,
  Count,
  Plus,
  MainPlus,
} from './ItemsCountHandler.styled';

const ItemsCountHandler = ({
  id,
  name,
  price,
  activeSize,
  activeType,
  typeNames,
  isFullScreen = false,
}: IItemsCountHandler) => {
  const dispatch = useAppDispatch();
  const namesCount = useAppSelector(selectItemsCount(id));
  const currentItemCount = useAppSelector(
    selectCurrentItemCount(name, activeSize, typeNames[activeType]),
  );

  const onClickAdd = () => {
    const item: ICartItem = {
      id,
      name,
      price,
      size: activeSize,
      type: typeNames[activeType],
      count: 0,
    };

    dispatch(addItem(item));
  };

  const onClickRemove = () => {
    if (currentItemCount < 1) return;

    const item: ICartItem = {
      id,
      name,
      price,
      size: activeSize,
      type: typeNames[activeType],
      count: 0,
    };

    dispatch(removeItem(item));
  };

  const handleAddKeyDownClick = (e: KeyboardEvent<SVGElement>) => {
    if (e.code === 'Enter') onClickAdd();
  };

  const handleRemoveKeyDownClick = (e: KeyboardEvent<SVGElement>) => {
    if (e.code === 'Enter') onClickRemove();
  };

  return namesCount ? (
    <Counter isFullScreen={isFullScreen}>
      <Minus
        onClick={onClickRemove}
        onKeyDown={(e) => handleRemoveKeyDownClick(e)}
        tabIndex={0}
        disabled={currentItemCount < 1}
        data-testid="itemsHandlerMinus"
      />
      <Count isFullScreen={isFullScreen}>{namesCount}</Count>
      <Plus
        onClick={onClickAdd}
        onKeyDown={(e) => handleAddKeyDownClick(e)}
        tabIndex={0}
        data-testid="itemsHandlerPlus"
      />
    </Counter>
  ) : (
    <ButtonAdd onClick={onClickAdd} isFullScreen={isFullScreen}>
      <MainPlus />
      <span>Add</span>
    </ButtonAdd>
  );
};

export default ItemsCountHandler;
