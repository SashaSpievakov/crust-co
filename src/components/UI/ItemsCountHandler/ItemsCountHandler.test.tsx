import '@testing-library/jest-dom';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import ItemsCountHandler from './ItemsCountHandler';
import rendererWithAllProviders from '../../../tests/helpers/rendererWithProviders';
import renderWithProviders from '../../../tests/helpers/renderWithProviders';
import { typeNames } from '../../FullItemBlock/FullItemBlock';
import { CartItemsMockProps } from '../../../tests/mocks/api/mockData';
import { IItemsCountHandler } from '../../../models/IItemsCountHandler';

const ItemsCountHandlerMockProps: IItemsCountHandler = {
  id: '6',
  name: 'Cheesburger Pizza',
  price: 7,
  activeSize: 1,
  activeType: 1,
  typeNames,
  isFullScreen: false,
};

describe('ItemsCountHandler Tests', () => {
  test('renders the ItemsCountHandler component', () => {
    const snapshot = rendererWithAllProviders(
      <ItemsCountHandler {...ItemsCountHandlerMockProps} />,
    );
    expect(snapshot).toMatchSnapshot();
  });

  test('renders the ItemsCountHandler component with items', () => {
    const snapshot = rendererWithAllProviders(
      <ItemsCountHandler {...ItemsCountHandlerMockProps} />,
      '/',
      { cart: CartItemsMockProps },
    );
    expect(snapshot).toMatchSnapshot();
  });

  describe('checks buttons clicks', () => {
    test('clicks add button', async () => {
      renderWithProviders(
        <ItemsCountHandler {...ItemsCountHandlerMockProps} />,
      );
      const addButton = screen.getByRole('button');

      userEvent.click(addButton);

      expect(await screen.findByText(/1/i)).toBeInTheDocument();
    });

    test('clicks minus button', async () => {
      renderWithProviders(
        <ItemsCountHandler {...ItemsCountHandlerMockProps} />,
      );
      const addButton = screen.getByRole('button');
      userEvent.click(addButton);

      const minusButton = await screen.findByTestId('itemsHandlerMinus');
      userEvent.click(minusButton);

      expect(screen.getByText(/add/i)).toBeInTheDocument();
    });

    test('clicks plus button', async () => {
      renderWithProviders(
        <ItemsCountHandler {...ItemsCountHandlerMockProps} />,
      );
      const addButton = screen.getByRole('button');
      userEvent.click(addButton);

      const plusButton = await screen.findByTestId('itemsHandlerPlus');
      userEvent.click(plusButton);
      userEvent.click(plusButton);
      userEvent.click(plusButton);
      userEvent.click(plusButton);

      expect(screen.getByText(/5/i)).toBeInTheDocument();
    });
  });
});
