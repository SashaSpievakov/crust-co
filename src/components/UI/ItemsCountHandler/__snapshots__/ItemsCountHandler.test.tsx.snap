// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemsCountHandler Tests renders the ItemsCountHandler component 1`] = `
Object {
  "children": Array [
    <svg
      className="sc-ipEyDJ fbIxLm"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 16 16"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 0a1 1 0 0 1 1 1v6h6a1 1 0 1 1 0 2H9v6a1 1 0 1 1-2 0V9H1a1 1 0 0 1 0-2h6V1a1 1 0 0 1 1-1z"
      />
    </svg>,
    <span>
      Add
    </span>,
  ],
  "props": Object {
    "className": "sc-bcXHqe sc-eDvSVe sc-gKPRtg fAfvZN fwWQTF gmBSoV",
    "onClick": [Function],
  },
  "type": "button",
}
`;

exports[`ItemsCountHandler Tests renders the ItemsCountHandler component with items 1`] = `
Object {
  "children": Array [
    <svg
      className="sc-pyfCe iRcGGc"
      data-testid="itemsHandlerMinus"
      disabled={true}
      fill="currentColor"
      height="1em"
      onClick={[Function]}
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 20 20"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clipRule="evenodd"
        d="M3 10a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1z"
        fillRule="evenodd"
      />
    </svg>,
    <span
      className="sc-kDvujY gwWHgA"
    >
      13
    </span>,
    <svg
      className="sc-jrcTuL qIHjZ"
      data-testid="itemsHandlerPlus"
      fill="currentColor"
      height="1em"
      onClick={[Function]}
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 20 20"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clipRule="evenodd"
        d="M10 3a1 1 0 011 1v5h5a1 1 0 110 2h-5v5a1 1 0 11-2 0v-5H4a1 1 0 110-2h5V4a1 1 0 011-1z"
        fillRule="evenodd"
      />
    </svg>,
  ],
  "props": Object {
    "className": "sc-ftTHYK kVxvuz",
  },
  "type": "section",
}
`;
